File: old/test/proto/cs/world.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Net;

/* This is the sub struct of "hello" */

namespace NodeProtocol {

	public class World {

		// public constants
		public const uint ID = 3;
		public const string NAME = "world";
		// public properties
		public string SenderUid { get; set; }
		public string[] Recipients { get; set; }
		public string Message { get; set; }
		public NodeProtocol.Sample1 Sample { get; set; }
		public uint Timestamp { get; set; }

		// private properties
		private byte[] _2bytes = new byte[2];
		private byte[] _4bytes = new byte[4];


		public World(byte[] buf = null) {
			if (buf != null) {
				int offset = 0;
				
				// unpack string as SenderUid
				Buffer.BlockCopy(buf, offset, _2bytes, 0, 2);
				Array.Reverse(_2bytes);
				offset += 2;
				int senderUidSize = (int)BitConverter.ToUInt16(_2bytes, 0);
				byte[] senderUidBytes = new byte[senderUidSize];
				Buffer.BlockCopy(buf, offset, senderUidBytes, 0, senderUidSize);
				offset += senderUidSize;
				SenderUid = Encoding.UTF8.GetString(senderUidBytes);
				// unpack recipients as array of string
				Buffer.BlockCopy(buf, offset, _2bytes, 0, 2);
				Array.Reverse(_2bytes);
				offset += 2;
				int recipientsLength = (int)BitConverter.ToUInt16(_2bytes, 0);
				Recipients = new string[recipientsLength];
				int recipientsIndex = 0;
				for (recipientsIndex = 0; recipientsIndex < recipientsLength; recipientsIndex++) {
					// unpack string as Recipients
					Buffer.BlockCopy(buf, offset, _2bytes, 0, 2);
					Array.Reverse(_2bytes);
					offset += 2;
					int recipientsSize = (int)BitConverter.ToUInt16(_2bytes, 0);
					byte[] recipientsBytes = new byte[recipientsSize];
					Buffer.BlockCopy(buf, offset, recipientsBytes, 0, recipientsSize);
					offset += recipientsSize;
					Recipients[recipientsIndex] = Encoding.UTF8.GetString(recipientsBytes);
				}
				// unpack string as Message
				Buffer.BlockCopy(buf, offset, _2bytes, 0, 2);
				Array.Reverse(_2bytes);
				offset += 2;
				int messageSize = (int)BitConverter.ToUInt16(_2bytes, 0);
				byte[] messageBytes = new byte[messageSize];
				Buffer.BlockCopy(buf, offset, messageBytes, 0, messageSize);
				offset += messageSize;
				Message = Encoding.UTF8.GetString(messageBytes);
				// Sample parameter: unpack Sample1 as Sample
				Buffer.BlockCopy(buf, offset, _2bytes, 0, 2);
				Array.Reverse(_2bytes);
				int sampleSize = (int)BitConverter.ToUInt16(_2bytes, 0);
				offset += 2;
				byte[] sampleBytes = new byte[sampleSize];
				Buffer.BlockCopy(buf, offset, sampleBytes, 0, sampleSize);
				Sample = new Sample1(sampleBytes);
				offset += sampleSize;
				// unpack Uint32 as Timestamp
				Buffer.BlockCopy(buf, offset, _4bytes, 0, 4);
				Array.Reverse(_4bytes);
				Timestamp = (uint)BitConverter.ToUInt32(_4bytes, 0);
				offset += 4;

			}
		}

		public byte[] Pack() {
			
			int totalBufSize = 0;
			// get byte size first: get byte size of string as SenderUid
			byte[] senderUidBytes = Encoding.UTF8.GetBytes(SenderUid);
			totalBufSize += 2 + senderUidBytes.Length;
			// Recipients is an array of string
			totalBufSize += 2;
			List<byte[]> recipientsPackByteList = new List<byte[]>();
			for (int recipientsIndex = 0, recipientsArrayLength = Recipients.Length; recipientsIndex < recipientsArrayLength; recipientsIndex++) {
				// get byte size first: get byte size of string as Recipients
				byte[] recipientsBytes = Encoding.UTF8.GetBytes(Recipients[recipientsIndex]);
				totalBufSize += 2 + recipientsBytes.Length;
				recipientsPackByteList.Add(recipientsBytes);
			}
			// get byte size first: get byte size of string as Message
			byte[] messageBytes = Encoding.UTF8.GetBytes(Message);
			totalBufSize += 2 + messageBytes.Length;
			// get byte size first: Sample parameter: get byte size of Sample1 as Sample
			/*** Sample ***/
			// get byte size first: Unique ID: get byte size of Uint32 as Sample.Id
			byte[] idBytes = BitConverter.GetBytes(Sample.Id);
			Array.Reverse(idBytes);
			totalBufSize += 4;
			// get byte size first: get byte size of string as Sample.Key
			byte[] keyBytes = Encoding.UTF8.GetBytes(Sample.Key);
			totalBufSize += 2 + keyBytes.Length;
			// get byte size first: Probably JSON: get byte size of string as Sample.Value
			byte[] valueBytes = Encoding.UTF8.GetBytes(Sample.Value);
			totalBufSize += 2 + valueBytes.Length;
			// get byte size first: get byte size of bool as Sample.Enabled
			int enabledBool = 0;
			if (Sample.Enabled == true) {
				enabledBool = 1;
			}
			byte[] enabledBytes = BitConverter.GetBytes(enabledBool);
			totalBufSize += 1;
			// Sample.Sample2list is an array of Sample2
			totalBufSize += 2;
			List<byte[]> sample2listPackByteList = new List<byte[]>();
			for (int sample2listIndex = 0, sample2listArrayLength = Sample.Sample2list.Length; sample2listIndex < sample2listArrayLength; sample2listIndex++) {
				// get byte size first: get byte size of Sample2 as Sample.Sample2list
				/*** Sample.Sample2list ***/
				// get byte size first: get byte size of string as Sample.Sample2list[sample2listIndex].Name
				byte[] nameBytes = Encoding.UTF8.GetBytes(Sample.Sample2list[sample2listIndex].Name);
				totalBufSize += 2 + nameBytes.Length;

				// create pack buffer
				byte[] buf = new byte[totalBufSize];
				int offset = 0;

				// pack string as Name
				_2bytes = BitConverter.GetBytes((ushort)nameBytes.Length);
				Array.Reverse(_2bytes);
				Buffer.BlockCopy(_2bytes, 0, buf, offset, 2);
				offset += 2;
				Buffer.BlockCopy(nameBytes, 0, buf, offset, nameBytes.Length);
				sample2listPackByteList.Add(sample2listBytes);
			}
			// get byte size first: get byte size of sbyte as Sample._eight
			byte[] _eightBytes = BitConverter.GetBytes(Sample._eight);
			totalBufSize += 1;
			// get byte size first: get byte size of short as Sample._sixteen
			byte[] _sixteenBytes = BitConverter.GetBytes(Sample._sixteen);
			totalBufSize += 2;
			// get byte size first: get byte size of int as Sample._thirtytwo
			byte[] _thirtytwoBytes = BitConverter.GetBytes(Sample._thirtytwo);
			totalBufSize += 4;
			// get byte size first: Date time object: get byte size of System.DateTime as Sample.Datetime
			double datetimeTime = (Sample.Datetime - new DateTime(1970, 1, 1)).TotalMilliseconds;
			byte[] datetimeBytes = BitConverter.GetBytes((double)datetimeTime);
			Array.Reverse(datetimeBytes);
			totalBufSize += 8;

			// create pack buffer
			byte[] buf = new byte[totalBufSize];
			int offset = 0;

			// Unique ID: pack Uint32 as Id
			Array.Reverse(idBytes);
			Buffer.BlockCopy(idBytes, 0, buf, offset, idBytes.Length);
			offset += 4;
			// pack string as Key
			_2bytes = BitConverter.GetBytes((ushort)keyBytes.Length);
			Array.Reverse(_2bytes);
			Buffer.BlockCopy(_2bytes, 0, buf, offset, 2);
			offset += 2;
			Buffer.BlockCopy(keyBytes, 0, buf, offset, keyBytes.Length);
			// Probably JSON: pack string as Value
			_2bytes = BitConverter.GetBytes((ushort)valueBytes.Length);
			Array.Reverse(_2bytes);
			Buffer.BlockCopy(_2bytes, 0, buf, offset, 2);
			offset += 2;
			Buffer.BlockCopy(valueBytes, 0, buf, offset, valueBytes.Length);
			// pack bool as Enabled
			Buffer.BlockCopy(enabledBytes, 0, buf, offset, enabledBytes.Length);
			offset += 1;
			// Sample.Sample2list is an array of Sample2
			_2bytes = BitConverter.GetBytes((ushort)Sample.Sample2list.Length);
			Array.Reverse(_2bytes);
			Buffer.BlockCopy(_2bytes, 0, buf, offset, 2);
			offset += 2;
			for (int sample2listIndex = 0, sample2listArrayLength = Sample.Sample2list.Length; sample2listIndex < sample2listArrayLength; sample2listIndex++) {
				// pack Sample2 as Sample2list

				int totalBufSize = 0;
				// get byte size first: get byte size of string as Name
				byte[] nameBytes = Encoding.UTF8.GetBytes(Name);
				totalBufSize += 2 + nameBytes.Length;

				// create pack buffer
				byte[] buf = new byte[totalBufSize];
				int offset = 0;

				// pack string as Name
				_2bytes = BitConverter.GetBytes((ushort)nameBytes.Length);
				Array.Reverse(_2bytes);
				Buffer.BlockCopy(_2bytes, 0, buf, offset, 2);
				offset += 2;
				Buffer.BlockCopy(nameBytes, 0, buf, offset, nameBytes.Length);
			}
			// pack sbyte as _eight
			Buffer.BlockCopy(_eightBytes, 0, buf, offset, _eightBytes.Length);
			offset += 1;
			// pack short as _sixteen
			Buffer.BlockCopy(_sixteenBytes, 0, buf, offset, _sixteenBytes.Length);
			offset += 2;
			// pack int as _thirtytwo
			Buffer.BlockCopy(_thirtytwoBytes, 0, buf, offset, _thirtytwoBytes.Length);
			offset += 4;
			// Date time object: pack System.DateTime as Datetime
			Array.Reverse(datetimeBytes);
			Buffer.BlockCopy(datetimeBytes, 0, buf, offset, datetimeBytes.Length);
			offset += 8;
			// get byte size first: get byte size of Uint32 as Timestamp
			byte[] timestampBytes = BitConverter.GetBytes(Timestamp);
			Array.Reverse(timestampBytes);
			totalBufSize += 4;

			// create pack buffer
			byte[] buf = new byte[totalBufSize];
			int offset = 0;

			// pack string as SenderUid
			_2bytes = BitConverter.GetBytes((ushort)senderUidBytes.Length);
			Array.Reverse(_2bytes);
			Buffer.BlockCopy(_2bytes, 0, buf, offset, 2);
			offset += 2;
			Buffer.BlockCopy(senderUidBytes, 0, buf, offset, senderUidBytes.Length);
			// Recipients is an array of string
			_2bytes = BitConverter.GetBytes((ushort)Recipients.Length);
			Array.Reverse(_2bytes);
			Buffer.BlockCopy(_2bytes, 0, buf, offset, 2);
			offset += 2;
			for (int recipientsIndex = 0, recipientsArrayLength = Recipients.Length; recipientsIndex < recipientsArrayLength; recipientsIndex++) {
				// pack string as Recipients
				_2bytes = BitConverter.GetBytes((ushort)recipientsPackByteList[recipientsIndex].Length);
				Array.Reverse(_2bytes);
				Buffer.BlockCopy(_2bytes, 0, buf, offset, 2);
				offset += 2;
				Buffer.BlockCopy(recipientsPackByteList[recipientsIndex], 0, buf, offset, recipientsPackByteList[recipientsIndex].Length);
				offset += recipientsPackByteList[recipientsIndex].Length;
			}
			// pack string as Message
			_2bytes = BitConverter.GetBytes((ushort)messageBytes.Length);
			Array.Reverse(_2bytes);
			Buffer.BlockCopy(_2bytes, 0, buf, offset, 2);
			offset += 2;
			Buffer.BlockCopy(messageBytes, 0, buf, offset, messageBytes.Length);
			// Sample parameter: pack Sample1 as Sample

			int totalBufSize = 0;
			// get byte size first: Unique ID: get byte size of Uint32 as Id
			byte[] idBytes = BitConverter.GetBytes(Id);
			Array.Reverse(idBytes);
			totalBufSize += 4;
			// get byte size first: get byte size of string as Key
			byte[] keyBytes = Encoding.UTF8.GetBytes(Key);
			totalBufSize += 2 + keyBytes.Length;
			// get byte size first: Probably JSON: get byte size of string as Value
			byte[] valueBytes = Encoding.UTF8.GetBytes(Value);
			totalBufSize += 2 + valueBytes.Length;
			// get byte size first: get byte size of bool as Enabled
			int enabledBool = 0;
			if (Enabled == true) {
				enabledBool = 1;
			}
			byte[] enabledBytes = BitConverter.GetBytes(enabledBool);
			totalBufSize += 1;
			// Sample2list is an array of Sample2
			totalBufSize += 2;
			List<byte[]> sample2listPackByteList = new List<byte[]>();
			for (int sample2listIndex = 0, sample2listArrayLength = Sample2list.Length; sample2listIndex < sample2listArrayLength; sample2listIndex++) {
				// get byte size first: get byte size of Sample2 as Sample2list
				/*** Sample2list ***/
				// get byte size first: get byte size of string as Sample2list[sample2listIndex].Name
				byte[] nameBytes = Encoding.UTF8.GetBytes(Sample2list[sample2listIndex].Name);
				totalBufSize += 2 + nameBytes.Length;

				// create pack buffer
				byte[] buf = new byte[totalBufSize];
				int offset = 0;

				// pack string as Name
				_2bytes = BitConverter.GetBytes((ushort)nameBytes.Length);
				Array.Reverse(_2bytes);
				Buffer.BlockCopy(_2bytes, 0, buf, offset, 2);
				offset += 2;
				Buffer.BlockCopy(nameBytes, 0, buf, offset, nameBytes.Length);
				sample2listPackByteList.Add(sample2listBytes);
			}
			// get byte size first: get byte size of sbyte as _eight
			byte[] _eightBytes = BitConverter.GetBytes(_eight);
			totalBufSize += 1;
			// get byte size first: get byte size of short as _sixteen
			byte[] _sixteenBytes = BitConverter.GetBytes(_sixteen);
			totalBufSize += 2;
			// get byte size first: get byte size of int as _thirtytwo
			byte[] _thirtytwoBytes = BitConverter.GetBytes(_thirtytwo);
			totalBufSize += 4;
			// get byte size first: Date time object: get byte size of System.DateTime as Datetime
			double datetimeTime = (Datetime - new DateTime(1970, 1, 1)).TotalMilliseconds;
			byte[] datetimeBytes = BitConverter.GetBytes((double)datetimeTime);
			Array.Reverse(datetimeBytes);
			totalBufSize += 8;

			// create pack buffer
			byte[] buf = new byte[totalBufSize];
			int offset = 0;

			// Unique ID: pack Uint32 as Id
			Array.Reverse(idBytes);
			Buffer.BlockCopy(idBytes, 0, buf, offset, idBytes.Length);
			offset += 4;
			// pack string as Key
			_2bytes = BitConverter.GetBytes((ushort)keyBytes.Length);
			Array.Reverse(_2bytes);
			Buffer.BlockCopy(_2bytes, 0, buf, offset, 2);
			offset += 2;
			Buffer.BlockCopy(keyBytes, 0, buf, offset, keyBytes.Length);
			// Probably JSON: pack string as Value
			_2bytes = BitConverter.GetBytes((ushort)valueBytes.Length);
			Array.Reverse(_2bytes);
			Buffer.BlockCopy(_2bytes, 0, buf, offset, 2);
			offset += 2;
			Buffer.BlockCopy(valueBytes, 0, buf, offset, valueBytes.Length);
			// pack bool as Enabled
			Buffer.BlockCopy(enabledBytes, 0, buf, offset, enabledBytes.Length);
			offset += 1;
			// Sample2list is an array of Sample2
			_2bytes = BitConverter.GetBytes((ushort)Sample2list.Length);
			Array.Reverse(_2bytes);
			Buffer.BlockCopy(_2bytes, 0, buf, offset, 2);
			offset += 2;
			for (int sample2listIndex = 0, sample2listArrayLength = Sample2list.Length; sample2listIndex < sample2listArrayLength; sample2listIndex++) {
				// pack Sample2 as Sample2list

				int totalBufSize = 0;
				// get byte size first: get byte size of string as Name
				byte[] nameBytes = Encoding.UTF8.GetBytes(Name);
				totalBufSize += 2 + nameBytes.Length;

				// create pack buffer
				byte[] buf = new byte[totalBufSize];
				int offset = 0;

				// pack string as Name
				_2bytes = BitConverter.GetBytes((ushort)nameBytes.Length);
				Array.Reverse(_2bytes);
				Buffer.BlockCopy(_2bytes, 0, buf, offset, 2);
				offset += 2;
				Buffer.BlockCopy(nameBytes, 0, buf, offset, nameBytes.Length);
			}
			// pack sbyte as _eight
			Buffer.BlockCopy(_eightBytes, 0, buf, offset, _eightBytes.Length);
			offset += 1;
			// pack short as _sixteen
			Buffer.BlockCopy(_sixteenBytes, 0, buf, offset, _sixteenBytes.Length);
			offset += 2;
			// pack int as _thirtytwo
			Buffer.BlockCopy(_thirtytwoBytes, 0, buf, offset, _thirtytwoBytes.Length);
			offset += 4;
			// Date time object: pack System.DateTime as Datetime
			Array.Reverse(datetimeBytes);
			Buffer.BlockCopy(datetimeBytes, 0, buf, offset, datetimeBytes.Length);
			offset += 8;
			// pack Uint32 as Timestamp
			Array.Reverse(timestampBytes);
			Buffer.BlockCopy(timestampBytes, 0, buf, offset, timestampBytes.Length);
			offset += 4;

			return buf;
		}

	}
}

File: old/test/proto/js/world.js
'use strict';

/* This is the sub struct of "hello" */

const id = 3;
const name = 'world';

module.exports.getId = getId;
module.exports.name = getName;
module.exports.create = create;
module.exports.pack = pack;
module.exports.unpack = unpack;

function getId() {
	return id;
}

function getName() {
	return name;
}

function create() {
	return {
		senderUid: '',
		recipients: [],
		message: '',
		sample: {
			id: 0,
			key: '',
			value: '',
			enabled: true,
			sample2list: [],
			_eight: 0,
			_sixteen: 0,
			_thirtytwo: 0,
			datetime: new Date()
		},
		timestamp: 0
	};
}

// converts an object into binary
function pack(obj) {
	const list = [];
	
	var senderUidBytes = Bin.from(obj.senderUid);
	var senderUidSize = senderUidBytes.length;
	var senderUidSizeBytes = Bin.alloc(2);
	senderUidSizeBytes.writeUInt16BE(senderUidSize, 0);
	list.push(senderUidSizeBytes);
	list.push(senderUidBytes);
	// .recipients is an array
	const recipientsLengthBytes = Bin.alloc(2);
	recipientsLengthBytes.writeUInt16BE(obj.recipients.length);
	list.push(recipientsLengthBytes);
	var recipientsIndex = 0;
	var recipientsLen = obj.recipients.length;
	for (var recipientsIndex = 0; recipientsIndex < recipientsLen; recipientsIndex++) {
		var recipientsArrBytes = Bin.from(obj.recipients[recipientsIndex]);
		var recipientsArrSize = recipientsArrBytes.length;
		var recipientsArrSizeBytes = Bin.alloc(2);
		recipientsArrSizeBytes.writeUInt16BE(recipientsArrSize, 0);
		list.push(recipientsArrSizeBytes);
		list.push(recipientsArrBytes);
	}
	var messageBytes = Bin.from(obj.message);
	var messageSize = messageBytes.length;
	var messageSizeBytes = Bin.alloc(2);
	messageSizeBytes.writeUInt16BE(messageSize, 0);
	list.push(messageSizeBytes);
	list.push(messageBytes);

	var sampleIdBytes = Bin.alloc(4);
	sampleIdBytes.writeUInt32BE(obj.sample.id, 0);
	list.push(sampleIdBytes);
	var sampleKeyBytes = Bin.from(obj.sample.key);
	var sampleKeySize = sampleKeyBytes.length;
	var sampleKeySizeBytes = Bin.alloc(2);
	sampleKeySizeBytes.writeUInt16BE(sampleKeySize, 0);
	list.push(sampleKeySizeBytes);
	list.push(sampleKeyBytes);
	var sampleValueBytes = Bin.from(obj.sample.value);
	var sampleValueSize = sampleValueBytes.length;
	var sampleValueSizeBytes = Bin.alloc(2);
	sampleValueSizeBytes.writeUInt16BE(sampleValueSize, 0);
	list.push(sampleValueSizeBytes);
	list.push(sampleValueBytes);
	var sampleEnabledBytes = Bin.alloc(1);
	sampleEnabledBytes.writeUInt8(obj.sample.enabled ? 0x01 : 0x00);
	list.push(sampleEnabledBytes);
	// .sample.sample2list is an array
	const sampleSample2listLengthBytes = Bin.alloc(2);
	sampleSample2listLengthBytes.writeUInt16BE(obj.sample.sample2list.length);
	list.push(sampleSample2listLengthBytes);
	var sampleSample2listIndex = 0;
	var sampleSample2listLen = obj.sample.sample2list.length;
	for (var sampleSample2listIndex = 0; sampleSample2listIndex < sampleSample2listLen; sampleSample2listIndex++) {

		var sampleSample2listNameBytes = Bin.from(obj.sample.sample2list[sampleSample2listIndex].name);
		var sampleSample2listNameSize = sampleSample2listNameBytes.length;
		var sampleSample2listNameSizeBytes = Bin.alloc(2);
		sampleSample2listNameSizeBytes.writeUInt16BE(sampleSample2listNameSize, 0);
		list.push(sampleSample2listNameSizeBytes);
		list.push(sampleSample2listNameBytes);
	}
	var sample_eightBytes = Bin.alloc(1);
	sample_eightBytes.writeInt8(obj.sample._eight, 0);
	list.push(sample_eightBytes);
	var sample_sixteenBytes = Bin.alloc(2);
	sample_sixteenBytes.writeInt16LE(obj.sample._sixteen, 0);
	list.push(sample_sixteenBytes);
	var sample_thirtytwoBytes = Bin.alloc(4);
	sample_thirtytwoBytes.writeInt32LE(obj.sample._thirtytwo, 0);
	list.push(sample_thirtytwoBytes);
	var sampleDatetimeBytes = Bin.alloc(8);
	sampleDatetimeBytes.writeDoubleBE(obj.sample.datetime.getTime(), 0);
	list.push(sampleDatetimeBytes);
	var timestampBytes = Bin.alloc(4);
	timestampBytes.writeUInt32BE(obj.timestamp, 0);
	list.push(timestampBytes);

	return Buffer.concat(list);
}

function unpack(buf) {
	const obj = {};
	
	var offset = 0;
	// unpack for obj.senderUid
	var senderUidSize = buf.readUInt16BE(offset);
	offset += 2;
	var senderUidBytes = Bin.alloc(senderUidSize);
	buf.copy(senderUidBytes, 0, offset);
	offset += senderUidSize;
	obj.senderUid = senderUidBytes.toString();
	// unpack for obj.recipients
	obj.recipients = [];
	var recipientsLength = buf.readUInt16BE(offset);
	offset += 2;
	var recipientsLen = recipientsLength;
	for (var recipientsIndex = 0; recipientsIndex < recipientsLength; recipientsIndex++) {
		var recipientsArrSize = buf.readUInt16BE(offset);
		offset += 2;
		var recipientsArrBytes = Bin.alloc(recipientsArrSize);
		buf.copy(recipientsArrBytes, 0, offset);
		offset += recipientsArrSize;
		obj.recipients.push(recipientsArrBytes.toString());
	}
	// unpack for obj.message
	var messageSize = buf.readUInt16BE(offset);
	offset += 2;
	var messageBytes = Bin.alloc(messageSize);
	buf.copy(messageBytes, 0, offset);
	offset += messageSize;
	obj.message = messageBytes.toString();
	// Sample parameter

	obj.sample = {};
	// Unique ID
	obj.sample.id = buf.readUInt32BE(offset);
	offset += 4;
	// unpack for obj.sample.key
	var sampleKeySize = buf.readUInt16BE(offset);
	offset += 2;
	var sampleKeyBytes = Bin.alloc(sampleKeySize);
	buf.copy(sampleKeyBytes, 0, offset);
	offset += sampleKeySize;
	obj.sample.key = sampleKeyBytes.toString();
	// Probably JSON
	var sampleValueSize = buf.readUInt16BE(offset);
	offset += 2;
	var sampleValueBytes = Bin.alloc(sampleValueSize);
	buf.copy(sampleValueBytes, 0, offset);
	offset += sampleValueSize;
	obj.sample.value = sampleValueBytes.toString();
	// unpack for obj.sample.enabled
	obj.sample.enabled = buf.readUInt8(offset) === 0x01 ? true : false;
	offset += 1;
	// unpack for obj.sample.sample2list
	obj.sample.sample2list = [];
	var sampleSample2listLength = buf.readUInt16BE(offset);
	offset += 2;
	var sampleSample2listLen = sampleSample2listLength;
	for (var sampleSample2listIndex = 0; sampleSample2listIndex < sampleSample2listLength; sampleSample2listIndex++) {

		obj.sample.sample2list[sampleSample2listIndex] = {};
		// unpack for obj.sample.sample2list.name
		var sampleSample2listNameSize = buf.readUInt16BE(offset);
		offset += 2;
		var sampleSample2listNameBytes = Bin.alloc(sampleSample2listNameSize);
		buf.copy(sampleSample2listNameBytes, 0, offset);
		offset += sampleSample2listNameSize;
		obj.sample.sample2list[sampleSample2listIndex].name = sampleSample2listNameBytes.toString();
	}
	// unpack for obj.sample._eight
	obj.sample._eight = buf.readInt8(offset);
	offset += 1;
	// unpack for obj.sample._sixteen
	obj.sample._sixteen = buf.readInt16LE(offset);
	offset += 2;
	// unpack for obj.sample._thirtytwo
	obj.sample._thirtytwo = buf.readInt32LE(offset);
	offset += 4;
	// Date time object
	obj.sample.datetime = new Date(buf.readDoubleBE(offset));
	offset += 8;
	// unpack for obj.timestamp
	obj.timestamp = buf.readUInt32BE(offset);
	offset += 4;

	return obj;
}

// Buffer handler static object Bin: Bin.from(<string>) and Bin.alloc(<size>)
const __methods = {
	from: null,
	alloc: null
};

getNodeVersion();

const Bin = {
	from: __methods.from,
	alloc: __methods.alloc
};

function getNodeVersion() {
	const vstring = process.version.replace('v', '');
	const v = parseInt(vstring.substring(0, vstring.indexOf('.')));
	if (v >= 7) {
		// version 7 or higher
		__methods.from = from;
		__methods.alloc = alloc;
	} else {
		__methods.from = oldFrom;
		__methods.alloc = oldAlloc;
	}
}

function from(str, enc) {
	return Buffer.from(str, enc);
}

function alloc(size) {
	return Buffer.alloc(size).fill(0);
}

function oldFrom(str, enc) {
	return new Buffer(str, enc);
}

function oldAlloc(size) {
	const buf = new Buffer(size);
	buf.fill(0);
	return buf;
}

